### PR TITLE
Feat/good reserve sdk

### PR DESCRIPTION
**Pull request by Ryjen1**
Resolves #34

This pull request introduces `@goodsdks/good-reserve`, a viem-based SDK for interacting with the GoodDollar Reserve to buy and sell G$. It includes React hook wrappers in `@goodsdks/react-hooks` and a complete demo application (`demo-reserve-swap`) demonstrating the full user flow.

## What was implemented
- **Core SDK (`@goodsdks/good-reserve`)**  
  Implements `GoodReserveSDK` which abstracts both Mento Broker routes (Celo production and XDC development) and Exchange Helper routes. Provides typed methods for quoting (`getBuyQuote` / `getSellQuote`), executing swaps (`buy` / `sell` with automatic approvals), and retrieving reserve diagnostics and transaction history.

- **Contract Mappings**  
  Inlines required ABIs and defines canonical `RESERVE_CONTRACT_ADDRESSES` for `production`, `staging`, and `development` environments across supported chains.  
  Addresses are pinned to GoodProtocol commit `04b5d250bc2802c7b04b7c4d87bc00222ea81e2b` (from `releases/deployment.json`).  
  ABIs are sourced from mento-core develop branch.

- **React Hooks Integration**  
  Added `useGoodReserve` hook in `@goodsdks/react-hooks`, consistent with existing identity-related hooks and exported through the package barrel.

- **Demo Application (`demo-reserve-swap`)**  
  Provides a working demonstration of the flow including real-time quotes, dynamic approval handling, G$ ↔ stable token swaps (cUSD on Celo, USDC on XDC dev), and decoded transaction history display.

## Refactoring & Improvements
- Parallel pool discovery in `getMentoExchangeId` now uses `Promise.all` to query multiple pools concurrently instead of sequentially.
- Replaced magic array indexing with named destructuring helpers (`extractPoolAddresses` and `extractPoolStats`) that align with viem's tuple/struct return shapes.
- Added optional `{ exactApproval: boolean }` setting in `GoodReserveSDKOptions`. Defaults to approving `maxUint256` for gas efficiency on repeated swaps; can be set to `true` to enforce exact-amount approvals.
- Expanded test suite to 23 tests, covering Mento Broker and Exchange Helper paths, approval logic, edge cases (zero amounts, unsupported chains), and event decoding.

## Checklist (DoD Verified)
- [x] `@goodsdks/good-reserve` workspace builds and exports public typed APIs from `src/index.ts`.
- [x] Local ABI definitions and local production/development contract-address maps are implemented from approved sources.
- [x] Public API/docs explicitly pin the ABI/address source references used for the implementation.
- [x] React hook wrapper(s) are added to `@goodsdks/react-hooks` and exported through package barrels.
- [x] One existing demo app includes a minimal reserve integration flow and is buildable.
- [x] Validation and safety guards exist for invalid amounts, chain/env mismatches, and execution constraints.
- [x] Tests or reproducible verification steps cover buy/sell execution and history/event decoding paths.
- [x] Lint, type-check, and affected workspace builds pass.
- [x] Included readmes

## How to Test
```bash
yarn install --immutable

yarn workspace @goodsdks/good-reserve build
yarn workspace @goodsdks/react-hooks build
yarn turbo run build --filter=demo-reserve-swap

yarn turbo run lint --filter=@goodsdks/good-reserve --filter=demo-reserve-swap
yarn tsc --noEmit

yarn workspace @goodsdks/good-reserve test

# Optional: run demo
yarn workspace demo-reserve-swap dev

### CODE REVIEW

**Review by blueogin — COMMENTED** [review]
Hey @Ryjen1 
It approves a very large amount. approving a large amount is not best practice.
I found the mapFriendlyError function in several files you changed - it looks redundant.

**Review by blueogin — DISMISSED** [review]
Incorrect buy/sell quote.

File: apps/demo-reserve-swap/src/components/useReserveSwapTx.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    gdDecimals: number
+  },
+): UseReserveSwapTxResult {
+  const { direction, stableToken, stableDecimals, gdDecimals } = params
+
+  const [txStatus, setTxStatus] = useState<TxStatus>("idle")
+  const [txResult, setTxResult] = useState<string | null>(null)
+  const [txError, setTxError] = useState<string | null>(null)
+
+  const execute = useCallback(
```
[the remark this answers is not in the document]
Comment (Ryjen1) [reply apps/demo-reserve-swap/src/components/useReserveSwapTx.ts]: I have also updated this on the latest fix. `ReserveSwap` now uses the boolean result returned from `executeSwap` instead of checking `txStatus`, so it no longer relies on stale async state.

**Review by L03TJ3 — CHANGES_REQUESTED** [review]
Mostly good, but as requested in the DoD
I am missing a demo flow that shows transaction history demonstrated/testable. (getTransactionHistory)

**Comment by Ryjen1** [comment]
Thanks for the note. I’ve  just added the demo flow that shows transaction history demonstrated and pushed the update.Please take another look.

@L03TJ3 @blueogin

**Review by L03TJ3 — APPROVED** [review]
Okay looks good and works! thanks for the contribution
